File: Dec1.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Scanner;

public class Dec1 {

  static void part1() {
    Scanner sc = new Scanner(System.in);
    var leftList = new ArrayList<Integer>();
    var rightList = new ArrayList<Integer>();
    while (sc.hasNext()) {
      int n = sc.nextInt();
      leftList.add(n);
      n = sc.nextInt();
      rightList.add(n);
    }
    Collections.sort(leftList);
    Collections.sort(rightList);
    long totalDist = 0;
    for (int i = 0; i < leftList.size(); i++) {
      totalDist += Math.abs(leftList.get(i) - rightList.get(i));
    }
    System.out.println(totalDist);
    sc.close();
  }

  static void part2() {
    Scanner sc = new Scanner(System.in);
    var leftList = new ArrayList<Integer>();
    var rightFreq = new HashMap<Integer, Integer>();
    while (sc.hasNext()) {
      int n = sc.nextInt();
      leftList.add(n);
      n = sc.nextInt();
      rightFreq.put(n, rightFreq.getOrDefault(n, 0) + 1);
    }
    int similarityScore = 0;
    for (var ele : leftList) {
      similarityScore += ele * rightFreq.getOrDefault(ele, 0);
    }
    System.out.println(similarityScore);
    sc.close();
  }

  public static void main(String[] args) {
    part2();
  }
}

File: aoc-2024-rs/src/dec1.rs
use std::{collections::HashMap, io::stdin, iter::zip};

fn part1() {
    let reader = stdin().lines();

    let mut left_col: Vec<i32> = Vec::new();
    let mut right_col: Vec<i32> = Vec::new();

    for line in reader {
        let line = line.unwrap();
        let mut split = line.trim().split(" ");

        let left_ele = split.next().unwrap().parse::<i32>().unwrap();
        left_col.push(left_ele);

        let right_ele = split.last().unwrap().parse::<i32>().unwrap();
        right_col.push(right_ele);
    }

    left_col.sort();
    right_col.sort();

    let total_dist: i32 = zip(left_col, right_col).map(|(a, b)| (a - b).abs()).sum();
    println!("{}", total_dist);
}

fn part2() {
    let reader = stdin().lines();

    let mut left_col: Vec<i32> = Vec::new();
    let mut right_freq: HashMap<i32, i32> = HashMap::new();

    for line in reader {
        let line = line.unwrap();
        let mut split = line.trim().split(" ");

        let left_ele = split.next().unwrap().parse::<i32>().unwrap();
        left_col.push(left_ele);

        let right_ele = split.last().unwrap().parse::<i32>().unwrap();
        if right_freq.contains_key(&right_ele) {
            right_freq.insert(right_ele, right_freq.get(&right_ele).unwrap() + 1);
        } else {
            right_freq.insert(right_ele, 1);
        }
    }

    let mut similarity_score = 0;
    for ele in left_col {
        if right_freq.contains_key(&ele) {
            similarity_score += ele * right_freq.get(&ele).unwrap();
        }
    }
    println!("{}", similarity_score);
}

pub fn main() {
    part2();
}
